### PR TITLE
Improvements to reference rules

### LINKED ISSUE #4281: Properties Windows for ProjectReferences experience is broken
When clicking on a project reference in Solution Explorer we show a few properties as editable in the properties window, but Private Assets, Include Assets and Exclude Assets and Reference Output Assembly error. The three Assets properties also are free text, rather than showing a good UI.

When setting Reference Output Assembly the node breaks but that is being tracked by a separate item.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #4281 and other bugs in rule files discovered after authoring tests.

There's a fair bit of churn in the tests due to consolidation. Ultimately the most important thing in this PR to review is that the changes to the rule files themselves are correct.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ProjectReference.xaml [hunk 1]
```diff
@@ -63,8 +63,7 @@
 
   <StringProperty Name="ExcludeAssets"
                   Description="Assets to exclude from this reference."
-                  DisplayName="Exclude Assets"
-                  Visible="True" />
+                  DisplayName="Exclude Assets" />
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ProjectReference.xaml]: Removed properties setters set to default.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedAssemblyReference.xaml [hunk 2]
```diff
@@ -135,7 +135,7 @@
     <BoolProperty.DataSource>
       <DataSource HasConfigurationCondition="False"
                   ItemType="Reference"
-                  Persistence="AssemblyReference"
+                  Persistence="ProjectFile"
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedAssemblyReference.xaml]: I think this property-level data source was copied from the rule-level one, and the `Persistence` property not updated.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedFrameworkReference.xaml [hunk 3]
```diff
@@ -8,7 +8,7 @@
   
   <Rule.DataSource>
     <DataSource HasConfigurationCondition="False"
-                ItemType="_ResolvedFrameworkReference"
+                ItemType="FrameworkReference"
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedFrameworkReference.xaml]: This should probably have been included in #5201. I'm not sure why this property is needed, as it doesn't effect anything as far as I can tell. However other such property-level `ItemType` properties match the outer item type, so I've updated this too for consistency.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedProjectReference.xaml [hunk 4]
```diff
@@ -40,7 +40,14 @@
 
   <BoolProperty Name="CopyLocalSatelliteAssemblies"
                 Description="Indicates whether the satellite assemblies of the reference target should be copied into this project's output directory."
-                DisplayName="Copy Local Satellite Assemblies" />
+                DisplayName="Copy Local Satellite Assemblies">
+    <BoolProperty.DataSource>
+      <DataSource HasConfigurationCondition="False"
+                  ItemType="ProjectReference"
+                  Persistence="ProjectFile"
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/ResolvedProjectReference.xaml]: This is an example of adding a property-level data source that persists to the project file on resolved dependencies. Without this, the rule-level data source is inherited and cannot write its value anywhere. There are several instances of this throughout the PR.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Items/Folder.xaml [hunk 5]
```diff
@@ -15,13 +15,13 @@
 
   <StringProperty Name="FolderName"
                   Category="Misc"
-                  Description="Name of this folder"
+                  Description="Name of this folder."
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Items/Folder.xaml]: Added trailing periods to descriptions. We already had a test for this, but it wasn't checking all rule files. This PR updates the tests to be more explicit about what rule files are included for each verification.

**Comment by davkean** [comment]
Didn't https://github.com/dotnet/project-system/issues/4281 mention providing a better UI for manipulating PrivateAssets/ExcludeAssets? I don't see it in this.

**Comment by drewnoakes** [comment]
I looked at that but the values are potentially many values (semicolon separated ), which `EnumProperty` doesn't support. `StringListProperty` could work here, but I don't see it adding much. Am I missing something?

**Pull request merged**